File: backend/models/resource.py
"""
    # Resource
    ```
        type = {
        title:string,
        file:string,
        description:string,
        documentType:string,
        documentFormat:string,
        username:string,
        hashtags:[string],
        subject: string ~ Subject id,
        course: string ~ Course id,
        createdAt: date
    }
    ```
"""

from typing_extensions import TypedDict

from backend import models


class Resource(TypedDict):
    title: str
    file: str
    description: str
    documentType: str
    documentFormat: str
    username: str
    hashtags: list[str]
    subjectId: str
    courseId: str

if __name__ == "__main__":
    obj = Resource({"title": "titulo", "file": "ficheiro.pdf", "tags": []})
    print(models.validate(obj))

File: frontend/models/Resource.ts
import { model, models, Schema } from 'mongoose';

interface IResource {
  title: string;
  description: string;
  documentTypeId: string;
  documentFormat: string;
  hashtags: string;
  subjectId: string;
  courseId: string;
  createdAt: Date;
  userEmail: string;
  favoritesNr: number;
  upvotesNr: number;
  downvotesNr: number;
  downloadsNr: number;
  edited: Date;
}

const ResourceSchema = new Schema<IResource>({
  title: { type: String, required: true },
  description: { type: String, required: true },
  documentTypeId: { type: String, ref: 'DocumentType', required: true },
  documentFormat: { type: String, required: true },
  hashtags: { type: String },
  subjectId: { type: String, ref: 'Subject', required: true },
  courseId: { type: String, ref: 'Course', required: true },
  createdAt: { type: Date, required: true },
  userEmail: { type: String, ref: 'User', required: true },
  favoritesNr: { type: Number, default: 0 },
  upvotesNr: { type: Number, default: 0 },
  downvotesNr: { type: Number, default: 0 },
  downloadsNr: { type: Number, default: 0 },
  edited: { type: Date },
});

const Resource =
  models.Resource || model<IResource>('Resource', ResourceSchema);

export default Resource;
